### PR TITLE
[foxy] Port CHOMP to MoveIt2

### PR DESCRIPTION
**Pull request by andreas-botbuilt**
### Description

This PR makes [CHOMP](http://docs.ros.org/en/melodic/api/moveit_tutorials/html/index.html) available in MoveIt2. CHOMP can be used as a stand-alone motion planner plugin (instead of OMPL) or as a post-processor for OMPL (where the seed trajectory for CHOMP is the output of OMPL).

### Changes
- CHOMP parameters are loaded from the namespace "chomp"
- Pose goals are converted to joint space goals (instead of being rejected)

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

### CODE REVIEW

File: moveit_planners/chomp/chomp_motion_planner/CMakeLists.txt [hunk 1]
```diff
@@ -1,42 +1,67 @@
-cmake_minimum_required(VERSION 3.1.3)
+cmake_minimum_required(VERSION 3.5)
 project(chomp_motion_planner)
 
-if(NOT "${CMAKE_CXX_STANDARD}")
+# Default to C99
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/CMakeLists.txt:4]: Also here, the defines should only be handled by `moveit_package()`

File: moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_parameters.h [hunk 2]
```diff
@@ -36,7 +36,7 @@
 
 #pragma once
 
-#include <ros/ros.h>
+#include "rclcpp/rclcpp.hpp"
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_parameters.h:39]: Looks like this include is not used at all. Feel free to remove

File: moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h [hunk 3]
```diff
@@ -52,5 +53,14 @@ class ChompPlanner
   bool solve(const planning_scene::PlanningSceneConstPtr& planning_scene,
              const planning_interface::MotionPlanRequest& req, const ChompParameters& params,
              planning_interface::MotionPlanDetailedResponse& res) const;
+
+  bool computePoseIK(const moveit::core::RobotModelConstPtr& robot_model, const std::string& group_name,
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h:57]: Why do these functions have to be added?

File: moveit_planners/chomp/chomp_motion_planner/package.xml [hunk 4]
```diff
@@ -1,20 +1,31 @@
...
+  <version>1.1.5</version>
   <description>chomp_motion_planner</description>
 
-  <version>1.1.5</version>
   <author email="gjones@willowgarage.com">Gil Jones</author>
   <author email="mail@mrinal.net">Mrinal Kalakrishnan</author>
-
   <maintainer email="chitt@live.in">Chittaranjan Srinivas Swaminathan</maintainer>
   <maintainer email="moveit_releasers@googlegroups.com">MoveIt Release Team</maintainer>
+  <maintainer email="andreas@botbuilt.com">atenpas</maintainer>
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/package.xml:12]: Are you interested in maintaining this package in the future?

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp [hunk 5]
```diff
@@ -35,18 +35,22 @@
...
 #include <eigen3/Eigen/Core>
+#include <eigen3/Eigen/LU>
 #include <random>
+#include <visualization_msgs/msg/marker_array.hpp>
+
+#include "rclcpp/rclcpp.hpp"
 
 namespace chomp
 {
+rclcpp::Logger LOGGER2 = rclcpp::get_logger("chomp_optimizer");
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:52]:
```suggestion
rclcpp::Logger LOGGER = rclcpp::get_logger("chomp_optimizer");
```
Why did you number the loggers?

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp [hunk 6]
```diff
@@ -67,22 +71,13 @@ ChompOptimizer::ChompOptimizer(ChompTrajectory* trajectory, const planning_scene
   , start_state_(start_state)
   , initialized_(false)
 {
-  std::vector<std::string> cd_names;
-  planning_scene->getCollisionDetectorNames(cd_names);
-
-  ROS_INFO_STREAM("The following collision detectors are available in the planning scene.");
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:73]: Why remove this? If it's too verbose, switching to DEBUG would be preferable

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp [hunk 7]
```diff
@@ -102,9 +97,11 @@ void ChompOptimizer::initialize()
   collision_detection::CollisionRequest req;
   collision_detection::CollisionResult res;
   req.group_name = planning_group_;
-  ros::WallTime wt = ros::WallTime::now();
+  // TODO
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:100]: Are you still planning on fixing the timers?

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp [hunk 8]
```diff
@@ -70,33 +78,89 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
 
   if (req.goal_constraints.size() != 1)
   {
-    ROS_ERROR_NAMED("chomp_planner", "Expecting exactly one goal constraint, got: %zd", req.goal_constraints.size());
+    RCLCPP_ERROR(LOGGER, "Expecting exactly one goal constraint, got: %zd", req.goal_constraints.size());
     res.error_code_.val = moveit_msgs::msg::MoveItErrorCodes::INVALID_GOAL_CONSTRAINTS;
     return false;
   }
 
+  // Convert to joint space goal.
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:86]: I still need to review this in detail. While this seems to be a really useful addition, I'd prefer if you could split this into a separate PR since it's unrelated to the migration work you have done

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp [hunk 9]
```diff
@@ -127,22 +191,22 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
...
   }
 
-  ROS_INFO_NAMED("chomp_planner", "CHOMP trajectory initialized using method: %s ",
-                 (params.trajectory_initialization_method_).c_str());
+  RCLCPP_INFO(LOGGER, "CHOMP trajectory initialized using method: %s ",
+              (params.trajectory_initialization_method_).c_str());
 
   // optimize!
-  ros::WallTime create_time = ros::WallTime::now();
+  std::chrono::time_point<std::chrono::system_clock> create_time = std::chrono::system_clock::now();
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp:209]: Using `chrono` directly is fine with me. `rclcpp::Duration` could make the code below a bit more readable and make it easier to access the seconds as double.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp [hunk 10]
```diff
@@ -34,9 +34,10 @@
 
 /* Author: Mrinal Kalakrishnan */
 
-#include <ros/ros.h>
 #include <chomp_motion_planner/chomp_trajectory.h>
 
+#include "rclcpp/rclcpp.hpp"
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp:39]: Also here, if there is no apparent change with this header, it can probably just be removed

File: moveit_planners/chomp/chomp_optimizer_adapter/CMakeLists.txt [hunk 11]
```diff
@@ -1,31 +1,56 @@
-cmake_minimum_required(VERSION 3.1.3)
+cmake_minimum_required(VERSION 3.10.2)
 project(moveit_chomp_optimizer_adapter)
 
-if(NOT "${CMAKE_CXX_STANDARD}")
+# Default to C99
+if(NOT CMAKE_C_STANDARD)
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_optimizer_adapter/CMakeLists.txt:5]: Just `moveit_package()` is enough here as well

File: moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp [hunk 12]
```diff
@@ -34,138 +34,138 @@
...
 #include <moveit/collision_distance_field/collision_detector_allocator_hybrid.h>
-
+#include <moveit/planning_interface/planning_interface.h>
+#include <moveit/planning_request_adapter/planning_request_adapter.h>
 #include <moveit/robot_state/conversions.h>
+#include <moveit/trajectory_processing/iterative_time_parameterization.h>
+#include <moveit/trajectory_processing/trajectory_tools.h>
 
-#include <vector>
+#include <class_loader/class_loader.hpp>
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp:46]: Better use `<pluginlib/class_loader.hpp>` (I know, it's not fully consistent in MoveIt, yet)
```suggestion
#include <pluginlib/class_loader.hpp>
```

File: moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp [hunk 13]
```diff
@@ -34,138 +34,138 @@
...
     }
-    if (!nh.getParam("max_iterations", params_.max_iterations_))
+    if (!node->get_parameter("chomp.max_iterations", params_.max_iterations_))
     {
       params_.max_iterations_ = 200;
-      ROS_INFO_STREAM("Param max_iterations was not set. Using default value: " << params_.max_iterations_);
+      RCLCPP_DEBUG(LOGGER, "Param max_iterations was not set. Using default value: %s", params_.max_iterations_);
     }
-    if (!nh.getParam("max_iterations_after_collision_free", params_.max_iterations_after_collision_free_))
+    if (!node->get_parameter("max_iterations_after_collision_free", params_.max_iterations_after_collision_free_))
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_optimizer_adapter/src/chomp_optimizer_adapter.cpp:76]: Did you leave out the prefix here intentionally?

**Review by henningkayser — COMMENTED** [review]
@andreas-botbuilt Thanks a lot, this is great work! This port came to a surprise for me. I left a fairly thorough review of the migration work, it would be great if you could split out the added support for pose constraints into a separate PR. I assume that you are working only on Foxy and don't run a `main` branch, correct? It would make it much easier for us if you could target these changes against `main` and we would backport them to `foxy`. If that's not possible, I'm sure we'll find another solution.

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp [hunk 14]
```diff
@@ -276,12 +273,12 @@ void ChompOptimizer::registerParents(const moveit::core::JointModel* model)
     {
       if (model->getParentLinkModel() == nullptr)
       {
-        ROS_ERROR_STREAM("Model " << model->getName() << " not root but has NULL link model parent");
+        RCLCPP_ERROR(LOGGER2, "Model %s not root but has NULL link model parent", model->getName());
```
Comment (henningkayser) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_optimizer.cpp:276]:
```suggestion
        RCLCPP_ERROR(LOGGER2, "Model %s not root but has NULL link model parent", model->getName().c_str());
```
Same in the ERROR below to make CI happy

File: moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp [hunk 15]
```diff
@@ -35,13 +35,17 @@
...
 #include <chomp_interface/chomp_planning_context.h>
-#include <moveit/trajectory_processing/iterative_time_parameterization.h>
 #include <moveit/robot_state/conversions.h>
+#include <moveit/trajectory_processing/iterative_time_parameterization.h>
+
+#include "rclcpp/rclcpp.hpp"
 
 namespace chomp_interface
 {
+rclcpp::Logger LOGGER3 = rclcpp::get_logger("chomp_optimizer");
```
[the remark this answers is not in the document]
Comment (andreas-botbuilt) [reply moveit_planners/chomp/chomp_interface/src/chomp_planning_context.cpp:45]: Changing this gives the following error: 

> /usr/bin/ld: CMakeFiles/moveit_chomp_interface.dir/src/chomp_planning_context.cpp.o:(.bss+0x0): multiple definition of `chomp_interface::LOGGER'; CMakeFiles/moveit_chomp_interface.dir/src/chomp_interface.cpp.o:(.bss+0x0): first defined here
> collect2: error: ld returned 1 exit status

**Comment by andreas-botbuilt** [comment]
> @andreas-botbuilt Thanks a lot, this is great work! This port came to a surprise for me. I left a fairly thorough review of the migration work, it would be great if you could split out the added support for pose constraints into a separate PR. I assume that you are working only on Foxy and don't run a `main` branch, correct? It would make it much easier for us if you could target these changes against `main` and we would backport them to `foxy`. If that's not possible, I'm sure we'll find another solution.

Thanks for your review. I'll try to make a new PR that targets these changes against `main`, so my most recent commit can be ignored, and I'll close this PR for now.
